src/Nethermind/Nethermind.Consensus/Processing/ParallelProcessing/MultiVersionMemory.cs

```diff
@@ -0,0 +1,194 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Consensus.Processing.ParallelProcessing;
+
+public class MultiVersionMemory<TLocation, TLogger>(ushort txCount, ParallelTrace<TLogger> parallelTrace) where TLogger : struct, IIsTracing where TLocation : notnull
```

Comment:
Another question is what Location should be?
We can have a map from Adress/StorageSlot to int and use that, or we can use complex types more directly.

Reply:
The ongoing question that I have (trying it out under #8482 partially) is would we benefit from having a concurrent map of Address/StorageCell -> int and then use it for navigation and addressing within a block. This additional mapping would incur some cost, but at the same time it would decrease the friction in all the underlying structures making either their keys atomic (for concurrent access) or simply smaller for hash sets etc. Having a uniform mapping would make everything generic of `<int>`.